src/Composer/Command/ShowCommand.php

```
@@ -168,13 +170,27 @@ protected function execute(InputInterface $input, OutputInterface $output)
             }
         } elseif ($input->getOption('all') && $composer) {
             $localRepo = $composer->getRepositoryManager()->getLocalRepository();
-            $installedRepo = new InstalledRepository(array($localRepo, $platformRepo));
+            $locker = $composer->getLocker();
+            if ($locker->isLocked()) {
+                $lockedRepo = $locker->getLockedRepository();
+                $installedRepo = new InstalledRepository(array($lockedRepo, $localRepo, $platformRepo));
+            } else {
+                $installedRepo = new InstalledRepository(array($localRepo, $platformRepo));
+            }
             $repos = new CompositeRepository(array_merge(array($installedRepo), $composer->getRepositoryManager()->getRepositories()));
         } elseif ($input->getOption('all')) {
             $defaultRepos = RepositoryFactory::defaultRepos($io);
             $io->writeError('No composer.json found in the current directory, showing available packages from ' . implode(', ', array_keys($defaultRepos)));
             $installedRepo = new InstalledRepository(array($platformRepo));
             $repos = new CompositeRepository(array_merge(array($installedRepo), $defaultRepos));
+        } elseif ($input->getOption('locked') && $composer) {
```

It should not check for $composer here IMO but rather below do `if (!$composer || !$composer->getLocker()->isLocked())`. Otherwise --locked will just silently be ignored which isn't great.

---

I'm not sure that I'm I understood correctly but I changed this.  Thanks for review

---

The point is that with the code the way you had it before, if I ran `composer show --locked` without composer.json it would just behave as if --locked was not passed, and show installed packages, which would be very confusing.

In reality though the getComposer() call at https://github.com/composer/composer/pull/8836/files#diff-b0c94b96dc86be178b130cd557412587R195 would have made it fail hard anyway as there was no composer.json, but I think this way is more explicit and not just working the intended way by chance.